### DISCUSSION TITLE
MeshLab 2021.05 is out!

### DISCUSSION THREAD

**Discussion by alemuntoni** [cnr-isti-vclab/meshlab#992] [General]
This new version brings lot of bugfixes and some new exciting features.

### Texture Map Defragmentation

We are happy to introduce a new "Texture Map Defragmentation" filter, which is the implementation of the paper [Texture Defragmentation for Photo-Reconstructed 3D Models](http://vcg.isti.cnr.it/Publications/2021/MCT21/).
Thanks to Andrea Maggiordomo for the implementation!!

### Loading files which contain more than one layer

MeshLab framework now allows to support file formats that may contain more than one mesh, and load them as multiple layers. For now, this feature is supported just into the 3DS importer, but we will introduce this possibility also to other file formats.

### Reintroduction of far clipping plane

Far clipping plane has been reintroduced (thanks to Christian Werner).

### Possibility to load custom shaders

MeshLab now allows to load custom GL shaders, in the GDP format.

### MeshLab Plugins

2021.05 version allows to load plugins at runtime, with a brand new plugin dialog and a more robust plugin management. In this context, we provide a [new github repository](https://github.com/cnr-isti-vclab/meshlab-extra-plugins) that will host some extra plugins supported by external developers, along with some other old plugins that have been dismissed from old MeshLab versions.

Since this is a brand new feature, we still have a small number of extra plugins, but we hope that the number will increase soon! Right now there are three extra plugins available:

- GlobalRegistration plugin, for rigid global alignment, developed by [Nicolas Mellado](https://www.irit.fr/recherches/STORM/MelladoNicolas/) and based on the [OpenGR](https://github.com/STORM-IRIT/OpenGR) library;
- VirtualGoniometer plugin, for computing angles across break edges, developed by [Jeff Calder](http://www-users.math.umn.edu/~jwcalder/) and based on the paper [The Virtual Goniometer: A new method for measuring angles on 3D models of fragmentary bone and lithics](https://arxiv.org/abs/2011.04898);
- The old "Surface Reconstruction: Poisson" filter that was in MeshLab before version 2016.12.

If you want to try them, you can go in the [Release Tab](https://github.com/cnr-isti-vclab/meshlab-extra-plugins/releases) of the extra-plugins repository, and download the archive depending on the OS you are running. Then unzip the archive, open MeshLab and go into Help -> Plugin Info -> Load Plugins! For now it is possible to install only plugins having the same version of MeshLab. We plan in the future to introduce compatibility between plugins and MeshLab versions.

If you are a developer and you are interested in sharing your algorithms/methods/implementations with the MeshLab community, what you need to do is just to develop a new MeshLab plugin! And consider that your plugin can be loaded at runtime also into [PyMeshLab](https://pymeshlab.readthedocs.io/en/latest/index.html) (free python bindings!).

To develop a MeshLab plugin is quite easy, and we are working to make it still easier! We are writing a documentation to explain how a MeshLab plugin is made, and how to build a new one. A first draft of the documentation can be found in the [Wiki page of the meshlab-extra-plugins repo](https://github.com/cnr-isti-vclab/meshlab-extra-plugins/wiki) (it is still under developement), and we also provide some [plugin examples](https://github.com/cnr-isti-vclab/meshlab-extra-plugins/tree/master/plugin_examples), which are good starting points to develop a MeshLab plugin!

The [meshlab-extra-plugin repository](https://github.com/cnr-isti-vclab/meshlab-extra-plugins) has been made specifically to collect and give visibility to plugins developed by external developers, therefore if you want to share your own MeshLab plugin, don't hesitate! We look forward to receive a pull request from you!

If you have any question, feel free to ask into this Discussions thread!

**Comment by endolith** [comment]
What is the difference between `MeshLab2021.05-windows.exe` and `MeshLab2021.05d-windows.exe`?

**Reply by alemuntoni** [reply]
2021.05d is built using double precision with scalar values.
Depending on the tasks you need to perform, you may prefer the double precision instead of the single precision on. 
It still is a beta version, but we are slowly improving its stability.
